### DISCUSSION TITLE
Circular dependency while using Joined Table inheritance

### DISCUSSION THREAD

**Discussion by crisog** [sqlalchemy/sqlalchemy#10388] [Usage Questions] [answered]
I'm pretty new to SQLAlchemy, so I'm coming here to see if there's obvious mistakes that others can signal but I (think) I've made some effort to figure it out but haven't nailed it yet.

I'm using Joined Table inheritance to abstract an "Account" from different user types. I'll go one by one, this is the account model which gets inherited by `User`
```python
class Account(UUIDAuditBase):
    __tablename__: str = "accounts"

    username: Mapped[str] = mapped_column(unique=True, nullable=False)
    email: Mapped[str] = mapped_column(unique=True, nullable=True)

    # Determines the type of account (i.e. user, provider, etc.)
    account_type: Mapped[str] = mapped_column(nullable=False)

    # -----------------
    # ORM Relationships
    # -----------------

    __mapper_args__ = {
        "polymorphic_on": account_type,
        "polymorphic_identity": "account",
    }

    # 1-M
    api_keys: Mapped[list["APIKey"]] = relationship(
        back_populates="account",
        lazy="joined",
        cascade="all, delete-orphan",
    )
```
The next model is going to be `User` which is what inherits `Account` to make up for the Joined Table inheritance
```python
class User(Account):
    __tablename__: str = "users"

    id: Mapped[UUID] = mapped_column(ForeignKey("accounts.id"), primary_key=True)
    # Omitting irrelevant fields (no relationships)
    
    # -----------------
    # ORM Relationships
    # -----------------

    __mapper_args__ = {"polymorphic_identity": "user"}

    # 1-1
    account: Mapped["Account"] = relationship(foreign_keys=[id], lazy="joined")

    # 1-M
    buckets: Mapped[list["Bucket"]] = relationship(
        back_populates="user",  # this means 'user' relationship in Bucket model
        lazy="joined",
        cascade="all, delete-orphan",
    )
```
The problem comes when I try to DELETE a user. The `api_keys` relationship on the account model has no issues and it seems to work with the cascade delete; the real issue is on the `buckets` relation. I'll share the `Bucket` model so you can visualize where the circular dependency generates.
```python
class Bucket(UUIDAuditBase):
    __tablename__: str = "buckets"
    user_id: Mapped[UUID] = mapped_column(ForeignKey("users.id"))

    # -----------------
    # ORM Relationships
    # -----------------

    # M-1
    user: Mapped["User"] = relationship(
        foreign_keys=[user_id],
        back_populates="buckets",  # back populates 'buckets' relationship in User
        lazy="joined",
    )

```
As you can see, there's a circular dependency between `User` and `Bucket` both referencing themselves in a relationship. However, this is me trying to be able to load the user buckets when I ask for an user information. And also getting the user's data if I ask for a bucket in my queries. The relationship in my head makes sense, but I might be just too naive on this one.

Ultimately, I just want to be able to keep my joined table inheritance, and keep the relation between user's owned buckets and the reference for the user in the bucket itself. Along with the ability to delete a user, and cascade delete the bucket if that happens.

The error message is the following:
```
sqlalchemy.exc.CircularDependencyError: Circular dependency detected. (DeleteState(<User at 0x7f2bf1a9e8d0>))
```

Any ideas?

**Comment by zzzeek** [comment]
Haven't tried yet but have you looked into post update ?  https://docs.sqlalchemy.org/en/20/orm/relationship_persistence.html#post-update.  That's usually what's needed if you are getting circular dependency at the instance level like that.

**Reply by crisog** [reply]
I tried it but ended up with a behavior that is not accurate to what I want. If I add `post_update=True` in the user relation at the buckets model, it tries to nullify the `user_id` field (I have it as non-nullable) and I want to cascade delete in all cases.

**Comment by CaselIT** [comment, marked as answer]
hi,

I'm not sure what's the point of this relationship:
```py
    # 1-1
    account: Mapped["Account"] = relationship(foreign_keys=[id], lazy="joined")
```

why you want an user to have itself set to also to the account property?

if I'm not mistaken this is always true:
```py
assert user is user.account
```

**Reply by crisog** [reply]
Removing this relation solved the circular dependency. It seems that I was looking the problem at a different angle.

Thank you :)

**Comment by zzzeek** [comment]
OK im back from vacation now, I would need full details to reproduce this since this is based on how you have the actual data set up, so I need the exact record structure you are persisting and then exactly what you are deleting, thanks!

**Reply by crisog** [reply]
Marked this as solved already - it seems that I was looking at the wrong place...

**Reply by zzzeek** [reply]
oh i see you have the FK as the PK in the primaryjoin, that's...strange I would think that would want to warn or something...
